definitions/apply-mapping/apply-mapping.yaml

```diff
@@ -0,0 +1,58 @@
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: apply-mapping
+spec:
+  params:
+    - name: applicationSnapshot
+      type: string
+      description: The ApplicationSnapshot in JSON format to apply the mapping to
+    - name: extraConfigPath
+      type: string
+      description: The path to the config file containing the mapping
+      default: ""
+  workspaces:
+    - name: config
+      description: The workspace where the extra config file containing the mapping resides
+  results:
+    - name: applicationSnapshot
+      description: |
+        The AppicationSnapshot in JSON format with the mapping applied. If there is no mapping key in the
+        extraConfigPath, this result will be the same as the original applicationSnapshot parameter.
+  steps:
+    - name: apply-mapping
+      image: quay.io/hacbs-release/release-utils
+      script: |
+        #!/usr/bin/env sh
+        set -eux
+
+        CONFIG_FILE="$(workspaces.config.path)/$(params.extraConfigPath)"
+        if [ ! -f "${CONFIG_FILE}" ] ; then
+            echo "No valid config file was provided."
+            echo '$(params.applicationSnapshot)' | tee $(results.applicationSnapshot.path)
+            exit 0
+        fi
+        if [[ $(yq '.mapping' "${CONFIG_FILE}") == "null" ]] ; then
+            echo "Config file contains no mapping key."
+            echo '$(params.applicationSnapshot)' | tee $(results.applicationSnapshot.path)
+            exit 0
+        fi
+
+        trap 'rm -f snapshot.yaml mapped_snapshot.yaml' EXIT
+
+        # Convert applicationSnapshot to YAML so it matches the config file
+        echo '$(params.applicationSnapshot)' | yq -P '.' > snapshot.yaml
+
+        # Merge the mapping key in the config file with the images key in the snapshot based on component name
```

Comment:
This does not remove unmapped components

Reply:
I tested the task in my cluster and yeah, it doesn't remove the unmapped components:
```
{
  "application": "m5-application",
  "images": [
    {
      "component": "component-1",
      "pullSpec": "quay.io/hacbs-release-tests/build-artifact-1@sha256:4ff3ca91275773af45cb4b0834e12b7eb47d1c18f770a0b151381cd227f4c253",
      "repository": "quay.io/hacbs-release-tests/component-1"
    },
    {
      "component": "component-2",
      "pullSpec": "quay.io/hacbs-release-tests/build-artifact-2@sha256:aa6c2c047467afc828e77e306041b7fa4a65734fe3449a54aa9c280822b0d87d"
    }
  ]
}
```
This is not valid as it would make the EC task to validate the unmapped component and that's something we don't want.

I haven't spent time on trying to fix the yq command, but a working jq query would look like this:
```
jq -s '.[0] as $snapshot | .[0].images + .[1].mapping | group_by(.component) | [.[] | select(length > 1)] | map(reduce .[] as $x ({}; . * $x)) as $mergedComponents | $snapshot | .images = $mergedComponents' snapshot.json mapping.json
```
Feel free to use it or ping me on Monday to work on the yq query together if you need help.